packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2822,26 +2822,56 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         // The call to fetch the snapshot is very expensive and not always needed.
         // It should only be done by the summarizerNode, if required.
         const snapshotTreeFetcher = async () => {
+            // When fetching from storage we will always get the latest version and do not use the ackHandle.
             const fetchResult = await this.fetchSnapshotFromStorage(
-                ackHandle,
-                summaryLogger,
+             null,
+             summaryLogger,
+             {
+                 eventName: "RefreshLatestSummaryGetSnapshot",
+                 ackHandle,
+                 summaryRefSeq,
+                 fetchLatest: true,
+             });
+             return fetchResult.snapshotTree;
+         };
+         const result = await this.summarizerNode.refreshLatestSummary(
+             proposalHandle,
+             summaryRefSeq,
+             snapshotTreeFetcher,
+             readAndParseBlob,
+             summaryLogger,
+         );
+
+         // In case we had to retrieve the latest snapshot and it is different than summaryRefSeq,
+         // wait for the delta manager to catch up.
+         if (result.latestSummaryUpdated === true && result.wasSummaryTracked === false) {
```

Comment:
It feels like we should do it before we call this.summarizerNode.refreshLatestSummary() above. Maybe it does not matter, but it looks safer - we refresh summary always with seq number that is in the past, not the future

Reply:
the problem is that we do not know if we will need to call fetch or not, the majority of the time we will not need to fetch,... Also , the behavior is exactly the same as when we call refreshLatestSummaryAckFromServer  and wait for the catch up after the refreshsummary.

Reply:
But is this relevant? If we are not going to fetch snapshot, that means (if I can read the code right) that this client submitted that summary, and thus it's definably ahead it will not wait for ops anyway.

Reply:
snapshotTreeFetcher  will only have been called and have the snapshot in specific cases (where refreshLatestSummaryFromSnapshot) .Are you saying we should return from the refreshsnapshot without calling refreshLatestSummaryFromSnapshot  until we had the deltamanager caught up ?

Reply:
Ah, I see it's not trivial. But here is what I'd do:
1. Break refreshLatestSummary() into two pieces - first part received 2 first args only and works only with pending summaries. If it's not a match, it returns some failure code.
2. Then this code calls getSnapshot() and waits for ops
3. Then it calls second part of refreshLatestSummary() - refreshLatestSummaryFromSnapshot() part (it needs to be public and added to interface).

Reply:
@NicholasCouri This is basically what my comment was. Let's continue the discussion here.
Vlad put it nicely - We refresh summary always with sequence number in the past. That is precisely my point. If the referenceSequenceNumber is in the future, it can lead to consistency problems.

Reply:
You can put the catchup logic in the `snapshotTreeFetcher` function right. So that if the summarizer node decides to download the snapshot, it will wait for catch up before proceeding.

Reply:
That is actually a great idea. I will re-evaluate the existing logic for submitSummary  as it  does refresh before catching up. It might be safer to do the same.

Reply:
@agarwal-navin see the latest iteration.